pvlib/iotools/bsrn.py

```diff
@@ -4,139 +4,449 @@
+    >>>     start=pd.Timestamp(2020,1,1), end=pd.Timestamp(2020,12,1),   # doctest: +SKIP
+    >>>     station='cab', username='yourusername', password='yourpassword')  # doctest: +SKIP
+
+    See Also
+    --------
+    pvlib.iotools.read_bsrn, pvlib.iotools.parse_bsrn
+
+    References
+    ----------
+    .. [1] `World Radiation Monitoring Center - Baseline Surface Radiation
+        Network (BSRN)
+        <https://bsrn.awi.de/>`_
+    .. [2] `BSRN Data Retrieval via FTP
+       <https://bsrn.awi.de/data/data-retrieval-via-ftp/>`_
+    .. [4] `BSRN Data Release Guidelines
+       <https://bsrn.awi.de/data/conditions-of-data-release/>`_
+    .. [3] `Update of the Technical Plan for BSRN Data Management, 2013,
+       Global Climate Observing System (GCOS) GCOS-174.
+       <https://bsrn.awi.de/fileadmin/user_upload/bsrn.awi.de/Publications/gcos-174.pdf>`_
+    """  # noqa: E501
+    # The FTP server uses lowercase station abbreviations
+    station = station.lower()
 
-    username and password are easily obtainable as described in the BSRN's
-    Data Release Guidelines [4]_.
+    # Generate list files to download based on start/end (SSSMMYY.dat.gz)
+    filenames = pd.date_range(start, end + pd.DateOffset(months=1), freq='1M')\
+        .strftime(f"{station}%m%y.dat.gz").tolist()
 
+    # Create FTP connection
+    with ftplib.FTP(BSRN_FTP_URL, username, password) as ftp:
+        # Change to station sub-directory (checks that the station exists)
+        try:
+            ftp.cwd(f'/{station}')
+        except ftplib.error_perm as e:
+            raise KeyError('Station sub-directory does not exist. Specified '
+                           'station is probably not a proper three letter '
+                           'station abbreviation.') from e
+        dfs = []  # Initialize list for monthly dataframes
+        non_existing_files = []  # Initilize list of files that were not found
+        for filename in filenames:
+            try:
+                bio = io.BytesIO()  # Initialize BytesIO object
+                # Retrieve binary file from server and write to BytesIO object
+                response = ftp.retrbinary(f'RETR {filename}', bio.write)
+                # Check that transfer was successfull
+                if not response.startswith('226 Transfer complete'):
+                    raise ftplib.Error(response)
+                # Save file locally if local_path is specified
+                if local_path is not None:
+                    # Create local file
+                    with open(os.path.join(local_path, filename), 'wb') as f:
+                        f.write(bio.getbuffer())  # Write local file
+                # Decompress/unzip and decode the binary file
+                text = gzip.decompress(bio.getvalue()).decode('latin1')
+                # Convert string to StringIO and parse data
+                dfi, metadata = parse_bsrn(io.StringIO(text), logical_records)

```

Comment:
There's a lot of shuffling data to/from buffers here (note that `gzip.decompress` wraps the input in a buffer as well).  I wonder if it might be cleaner to use `gzip.GzipFile` directly, maybe with an `io.TextIOWrapper`?  Maybe not, and probably this is a small fraction of total runtime anyway, so maybe not worth optimizing.

Reply:
I took a lot of fiddling around to find this solution, as I was stumbling into errors for all the other solutions I investigated.

When I run `gzip.GzipFile(bio)`, I get the following error:
`TypeError: expected str, bytes or os.PathLike object, not _io.BytesIO`

Reply:
Right, it would use the `fileobj` parameter of `GzipFile`, I think this might work:

```python
bio.seek(0)
gzip_file = io.TextIOWrapper(gzip.GzipFile(fileobj=bio), encoding='utf-8')  # or latin1?
dfi, metadata = parse_bsrn(gzip_file, logical_records)
```

Again, not sure this is worth worrying about though.

Reply:
My solution takes 83 ms, whereas yours take 20.6 µs. So certainly a bit of a performance gain :)

I'll change it to your solution